doc: add core concepts of components (#420)
karmada-io/website: docs/core-concepts/components.md

---
title: Karmada Components
---

This document outlines the various components you need to have for a complete and working Karmada.

// 此处待补图完善

## Control Plane Components

### karmada-apiserver

This component uses the official image of Kuberntes, responsible for handling API requests, and communicating with the karmada-aggregated-apiserver, various controllers, karmada-scheduler, karmada-webhook, and persisting data to etcd.

### karmada-aggregated-apiserver

This component is used to access resources on the member cluster through karmada-apiserver, achieving unified authentication on the Karmada control plane and authentication through user impersonation.

The resources installed in the aggregate apiserver include clusters, clusters/status, clusters/proxy.

### kube-controller-manager

This component uses the official image of Kuberntes, the difference is that only several controllers are enabled such as namespace, garbagecollector, serviceaccount-token controllers.

> > When users submit Deployment or other k8s standard resources to karmada-apiserver, they are only recorded in the etcd of the karmada control plane and subsequently synchronized to the member cluster. These Deployment resources do not undergo a reconciliation (such as creating a pod) in the karmada control plane cluster.

Reviewer: ```suggestion
> When users submit Deployment or other Kubernetes standard resources to the karmada-apiserver,
> they are solely recorded in the etcd of the Karmada control plane. Subsequently, these resources are
> synchronized with the member cluster. However, it's important to note that these Deployment resources
> do not undergo reconciliation processes (such as pod creation) in the Karmada control plane cluster.
```

3. Binding Controller: watches ResourceBinding objects and create a Work object corresponding to each cluster with a single resource manifest.
4. Execution Controller: watches Work objects. When Work objects are created, the controller will distribute the resources to member clusters.

### karmada-scheduler

This component is responsible for scheduling k8s native API resource objects (including CRD resources) to member clusters, including scheduling strategies and timing. At the same time, it also supports extending scheduling capabilities through plugins.

For some resource objects that need to calculate the replica ratio, the scheduling plugin will first filter out some clusters, and then make a decision on the number of replicas that the resource object should distribute in each cluster.

### karmada-webhook

This component is responsible for verifying the format of resources object which is in the charge of , as well as intercepting and modifying relevant data before it entering the ETCD.

### etcd

Consistent and highly-available key value store used as Karmada' backing store for all karmada API objects. 

## Data Plane Components

> ### karmada-agent

Reviewer 2: I think `karmada-agent` also belongs to the control plane.

To address this issue, the estimator component was introduced, which calculates the number of callable replicas for each node based on resource requests, thereby calculating the true number of schedulable replicas for a cluster.

### karmada-descheduler

This component is a cronjob responsible for detecting all Deployments at regular intervals (every 2 minutes by default), and triggering reschedule based on instance state changes in member clusters. The descheduler only takes effect when the scheduling strategy is dynamic division and it perceives how many instance state has changed by calling scheduler-estimator.

### karmada-search

Karmada search is used to cache resource objects and events deployed in member clusters, and to provide retrieval services externally through search APIs.

Unlike the query services provided by the aggregation API, cache retrieval does not require accessing the target cluster and is targeted at scenarios that require frequent data search and analysis, such as building real-time resources and event dashboards.

> ### karmadactl / kubectl-karmada

Reviewer 3: It might be better to publicly introduce these two tasks separately, as they have different usage methods from the perspective of user experience.

---
title: Karmada Components
---

This document provides an overview of the components required for a fully functional and operational Karmada setup.

![components](../resources/general/components.png)

## Control Plane Components

The control plane's components make global decisions about the cluster (for example, scheduling),

> as well as detecting and responding to cluster events.

Reviewer 2: I'm sorry for I gave the wrong link of `Cluster Registration Mode`, 
`https://kubernetes.io/docs/concepts/overview/components/` --> `https://karmada.io/docs/next/userguide/clustermanager/cluster-registration`.

### karmada-apiserver

This component uses the official image of Kubernetes and plays a crucial role in handling API requests,
facilitating communication with the karmada-aggregated-apiserver, various controllers, karmada-scheduler,

> karmada-webhook, and persisting data in etcd.

Reviewer 2: ```
The API server is a component of the Karmada control plane that exposes the Karmada API in addition to the Kubernetes API. The API server is the front end of the Karmada control plane.

Karmada API server directly uses the implementation of `kube-apiserver` from Kubernetes, which is the reason why Karmada is naturally compatible with Kubernetes API. That makes integration with the Kubernetes ecosystem very simple for Karmada, such as allowing users to use `kubectl` to operate Karmada, [integrating with ArgoCD](https://karmada.io/docs/next/userguide/cicd/working-with-argocd), [integrating with Flux](https://karmada.io/docs/next/userguide/cicd/working-with-flux), and so on.
```

---
title: Components
---

This document provides an overview of the components required for a fully functional and operational Karmada setup.

![components](../resources/general/components.png)

## Control Plane Components

A complete and working Karmada control plane consists of the following components. 
The `karmada-agent` might be optional, that depends on 
[Cluster Registration Mode](https://kubernetes.io/docs/concepts/overview/components/).

### karmada-apiserver

The API server is a component of the Karmada control plane that exposes the Karmada API in addition to the Kubernetes API. 
The API server is the front end of the Karmada control plane.

> Karmada API server directly uses the implementation of `kube-apiserver` from Kubernetes, which is the reason why Karmada 

Reviewer 3: If we directly state here that we are using the LTS version of kube-apiserver, wouldn't it be more explicit?

is naturally compatible with Kubernetes API. That makes integration with the Kubernetes ecosystem very simple for Karmada, 
such as allowing users to use `kubectl` to operate Karmada, 
[integrating with ArgoCD](https://karmada.io/docs/next/userguide/cicd/working-with-argocd), 
[integrating with Flux](https://karmada.io/docs/next/userguide/cicd/working-with-flux), and so on.

### karmada-aggregated-apiserver

This component is used to access resources on the member cluster through karmada-apiserver,
achieving unified authentication on the Karmada control plane and authentication through user impersonation.

> The resources installed in the aggregate apiserver include clusters, clusters/status, clusters/proxy.

Reviewer 2: I'd like to give a high-lever overview of this component, something like:
```
The aggregate API server is an extended API server implemented using [Kubernetes API Aggregation Layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) technology.
It offers [Cluster API](https://github.com/karmada-io/karmada/blob/master/pkg/apis/cluster/types.go) and related sub-resources, such as `cluster/status` and `cluster/proxy`, it implements advanced capabilities like [Aggregated Kubernetes API](https://karmada.io/docs/userguide/globalview/aggregated-api-endpoint) which can be used to access member clusters through Karmada API server.
```

### kube-controller-manager

This component uses the official image of Kuberntes, the difference is that only several controllers are

> enabled such as namespace, garbagecollector, serviceaccount-token controllers.

Reviewer 3: Can we directly attach the relevant instructions to ?

Author: done

> When users submit Deployment or other Kubernetes standard resources to the karmada-apiserver,
> they are solely recorded in the etcd of the Karmada control plane. Subsequently, these resources are
> synchronized with the member cluster. However, it's important to note that these Deployment resources
> do not undergo reconciliation processes (such as pod creation) in the Karmada control plane cluster.

### karmada-controller-manager

This component performs operations based on the API objects you created through the karmada-apiserver

This component runs various controllers, which watch karmada objects and then talk to the underlying
clusters' API servers to create regular Kubernetes resources. Some types of these controllers are:

1. Cluster Controller: attaches kubernetes clusters to Karmada for managing the lifecycle of the clusters
   by creating cluster objects.

> 2. Policy Controller: watches PropagationPolicy objects. When a PropagationPolicy object is added, the

Author: keep the same as [Architecture](https://karmada.io/docs/next/core-concepts/architecture)

> 2. Policy Controller: watches PropagationPolicy objects. When a PropagationPolicy object is added, the

Reviewer 2: I have the same feeling, please don't list the examples here, please link to https://karmada.io/docs/administrator/configuration/configure-controllers/#karmada-controllers

Besides, the agent also in charge of collecting and reporting current cluster status to control plane.

## Addons

### karmada-scheduler-estimator

This component is responsible for providing the karmada-scheduler with a more precise desired state of the running instances of the member cluster.
Early karmada-scheduler only supported scheduling the number of replicas based on the total number of cluster resources.
In this case, scheduling failure occurred when the total cluster resources were sufficient but each node resources were insufficient.

To address this issue, the estimator component was introduced, which calculates the number of callable replicas for
each node based on resource requests, thereby calculating the true number of schedulable replicas for a cluster.

### karmada-descheduler

This component is a cronjob responsible for detecting all Deployments at regular intervals (every 2 minutes by default),
and triggering reschedule based on instance state changes in member clusters.

The descheduler only takes effect when the scheduling strategy is dynamic division
and it perceives how many instance state has changed by calling scheduler-estimator.

> ### karmada-search

Reviewer 3: This part of the functionality is also carried by Karmada Search, which can be further explained.

https://karmada.io/docs/userguide/globalview/proxy-global-resource

---
title: Components
---

This document provides an overview of the components required for a fully functional and operational Karmada setup.

![components](../resources/general/components.png)

## Control Plane Components

A complete and working Karmada control plane consists of the following components. 
The `karmada-agent` might be optional, that depends on 

> [Cluster Registration Mode](https://karmada.io/docs/next/userguide/clustermanager/cluster-registration).

Reviewer: ```suggestion
[Cluster Registration Mode](../userguide/clustermanager/cluster-registration).
```
How about a relative url?

> [Cluster Registration Mode](https://karmada.io/docs/next/userguide/clustermanager/cluster-registration).

Reviewer 4: ```suggestion
[Cluster Registration Mode](../userguide/clustermanager/cluster-registration).
```
Can you use relative paths to reference other documents?

### karmada-apiserver

The API server is a component of the Karmada control plane that exposes the Karmada API in addition to the Kubernetes API. 
The API server is the front end of the Karmada control plane.

Karmada API server directly uses the implementation of `kube-apiserver` from Kubernetes, which is the reason why Karmada 
is naturally compatible with Kubernetes API. That makes integration with the Kubernetes ecosystem very simple for Karmada, 
such as allowing users to use `kubectl` to operate Karmada, 

> [integrating with ArgoCD](https://karmada.io/docs/next/userguide/cicd/working-with-argocd), 

Reviewer 4: Ditto

> [integrating with Flux](https://karmada.io/docs/next/userguide/cicd/working-with-flux) and so on.

Reviewer 4: Ditto

### karmada-aggregated-apiserver

The aggregate API server is an extended API server implemented using
[Kubernetes API Aggregation Layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) technology.
It offers [Cluster API](https://github.com/karmada-io/karmada/blob/master/pkg/apis/cluster/types.go)
and related sub-resources, such as `cluster/status` and `cluster/proxy`, it implements advanced capabilities like

> [Aggregated Kubernetes API](https://karmada.io/docs/userguide/globalview/aggregated-api-endpoint)

Reviewer 4: Ditto

which can be used to access member clusters through karmada-apiserver.

### kube-controller-manager

The kube-controller-manager is composed of a bunch of controllers.
Karmada just inherits some controllers from the official image of Kubernetes to keep a consistent user experience and behavior.

It's worth noting that not all controllers are needed by Karmada, for the recommended controllers please refer to

> [Recommended Controllers](https://karmada.io/docs/administrator/configuration/configure-controllers#required-controllers).

Reviewer 4: Ditto

> Note: When users submit Deployment or other Kubernetes standard resources to the karmada-apiserver,
> they are solely recorded in the etcd of the Karmada control plane. Subsequently, these resources are
> synchronized with the member cluster. However, these Deployment resources
> do not undergo reconciliation processes (such as pod creation) in the Karmada control plane cluster.

### karmada-controller-manager

The karmada-controller-manager runs various custom controller processes.

The controllers watch Karmada objects and then talk to the underlying clusters' API servers to create regular Kubernetes resources.

> The controllers are listed at [Karmada Controllers](https://karmada.io/docs/administrator/configuration/configure-controllers/#karmada-controllers)

Reviewer 5: ```suggestion
The controllers are listed at [Karmada Controllers](https://karmada.io/docs/administrator/configuration/configure-controllers/#karmada-controllers).
```

The scheduler then ranks each valid cluster and binds the resource to the most suitable cluster.

### karmada-webhook

karmada-webhooks are HTTP callbacks that receive Karmada/Kubernetes API requests and do something with them.
You can define two types of karmada-webhook, validating webhook and mutating webhook.

Mutating webhooks are invoked first, and can modify objects sent to the karmada-apiserver to enforce custom defaults.
After all object modifications are complete, and after the incoming object is validated by the karmada-apiserver,
validating webhooks are invoked and can reject requests to enforce custom policies.

> Note: Webhooks that need to guarantee they see the final state of the object in order to enforce policy should use a validating webhook,
> since objects can be modified after being seen by mutating webhooks.

### etcd

Consistent and highly-available key value store used as Karmada' backing store for all Karmada/Kubernetes API objects.

If your Karmada uses etcd as its backing store, make sure you have a back up plan for the data.

You can find in-depth information about etcd in the official [documentation](https://etcd.io/docs/).

### karmada-agent

> Karmada has two [Cluster Registration Mode](https://karmada.io/docs/next/userguide/clustermanager/cluster-registration) such as Push and Pull,

Reviewer 4: Ditto

> each node based on resource requests, thereby calculating the true number of schedulable replicas for a cluster.

### karmada-descheduler

This component is responsible for detecting all replicas at regular intervals (two minutes by default),
and triggering reschedule based on instance state changes in member clusters.

The descheduler only takes effect when the scheduling strategy is dynamic division
and it perceives how many instance state has changed by calling scheduler-estimator.

### karmada-search

The karmada-search starts an aggregated server. It provides capabilities such as global search and resource proxy in a multi-cloud environment.

The ability of [global search](https://karmada.io/docs/tutorials/karmada-search/) is to cache resource objects and events across multiple clusters,
and to provide graphical retrieval services externally through search APIs.

The ability of [resource proxy](https://karmada.io/docs/userguide/globalview/proxy-global-resource/) allows users to access all the resources both in karmada controller panel and member clusters.

### karmadactl

> Users can use karmadactl as client to enjoy karmada, such as performing join/unjoin of a member cluster,

Reviewer 5: ```suggestion
Users can use karmadactl as client to enjoy Karmada, such as performing join/unjoin of a member cluster,
```

---
title: Karmada Components
---

This document provides an overview of the components required for a fully functional and operational Karmada setup.

![components](../resources/general/components.png)

## Essential Components

### karmada-apiserver

> This component uses the official image of Kubernetes and plays a crucial role in handling API requests,
> 
> facilitating communication with the karmada-aggregated-apiserver, various controllers, karmada-scheduler,
> 
> karmada-webhook, and persisting data in etcd.

Reviewer: ```suggestion
This component uses the official image of Kubernetes and plays a crucial role in handling API requests,
facilitating communication with the karmada-aggregated-apiserver, various controllers, karmada-scheduler,
karmada-webhook, and persisting data in etcd.
```
Do not add any blank line if it is a para or a sentence.

Author: Sorry, the blank line is automatic generated by my editor `typora`, I didn't pay attention to this. I'll fix it.
